src/highdicom/seg/sop.py

```diff
@@ -2572,42 +2737,52 @@ def get_pixels_by_source_frame(
-                        'parameter.'
-                    )
-                    raise ValueError(msg)
+                msg = (
+                    f'Source frame number {src_frm_num} does not '
+                    'match any referenced source frame. To return '
+                    'an empty segmentation mask in this situation, '
+                    "use the 'assert_missing_frames_are_empty' "
+                    'parameter.'
+                )
+                raise ValueError(msg)
 
-            # Iterate over segment numbers for this source frame
-            for c, seg_num in enumerate(segment_numbers):
-                # Use LUT to find the segmentation frame containing
-                # the source frame and segment number
-                qry = np.array([src_frm_num, seg_num])
-                seg_frm_indices = np.where(np.all(lut == qry, axis=1))[0]
-                if len(seg_frm_indices) == 1:
-                    seg_frames[r, c] = seg_frm_indices[0] + 1
-                elif len(seg_frm_indices) > 0:
-                    # This should never happen due to earlier checks
-                    # on unique rows
-                    raise RuntimeError(
-                        'An unexpected error was encountered during '
-                        'indexing of the segmentation image. Please '
-                        'file a bug report with the highdicom repository.'
-                    )
-                # else no segmentation frame found for this segment number,
-                # assume this frame is empty and leave the entry in seg_frames
-                # as -1
+        source_frame_numbers_arr = np.array(source_frame_numbers, np.uint)
+
+        # Loop through each row of the LUT and determine whether the
+        # corresponding segmentation frame is required in the output, and if so,
+        # where
+        for seg_frm, lut_row in enumerate(lut):
+            # Source frame index and segment number for this LUT row
+            lut_src, lut_segno = lut_row.ravel()
+
+            # Find indices (if any) of these value in the lists of required
+            # output source frames and segments
+            out_frm_inds = np.where(source_frame_numbers_arr == lut_src)[0]
+            out_seg_inds = np.where(segment_numbers == lut_segno)[0]
+
+            if len(out_frm_inds) > 1:
+                # This should never happen due to earlier checks
+                # on unique rows
+                raise RuntimeError(
+                    'An unexpected error was encountered during '
+                    'indexing of the segmentation image. Please '
+                    'file a bug report with the highdicom repository.'
+                )
+
+            # Frame is only required if there is an index for both
+            if len(out_frm_inds) > 0 and len(out_seg_inds) > 0:
+                r = out_frm_inds[0]
+                c = out_seg_inds[0]
+                seg_frames[r, c] = seg_frm + 1
```

Comment:
This appears to be a duplication. Can we factor this out into a function?

Reply:
See https://github.com/herrmannlab/highdicom/pull/208#issuecomment-1358495069